### PR TITLE
feat: allow users to join a drop, assign a random NFT to user at drop…

### PR DESCRIPTION
**Pull request by BerinikeTech**
… time, let user only buy the assigned NFT

### CODE REVIEW

File: blockchain/contracts/HelloWorld.sol [hunk 1]
```diff
@@ -1,52 +1,124 @@
...
+    ) public {
+        uint256 nftHash = generateRandomNumber(_number);
         NFTOwnership memory nftOwnership;
         nftOwnership.nftId = nftHash;
         nftOwnership.owner = nftOwner;
-        nftOwnership.droptime = _droptime;
+        nftOwnership.reservedFor = nftOwner;
+        nftOwnership.dropTime = _dropTime;
+        nftOwnership.dropHash = _dropHash;
         nftOwnerships[nftHash] = nftOwnership;
```
Comment (JaPilgrim) [inline blockchain/contracts/HelloWorld.sol:55]: when initializing, nftHash is used as Index

Comment (JaPilgrim) [inline blockchain/contracts/HelloWorld.sol:55]: when initializing, nftHash is used as Index

File: blockchain/contracts/HelloWorld.sol [hunk 2]
```diff
@@ -1,52 +1,124 @@
...
+        require(joinedUsers.size() <= 10, "You cannot join the drop anymore.");
+        joinedUsers.pushAddress(msg.sender);
     }
 
-    function buyNFT(uint256 _price, uint256 _nftHash) public payable  {
-        require(nftOwnerships[_nftHash].droptime<=block.timestamp,"Droptime not yet reached!");
-        //string(abi.encodePacked("Drop has not started yet! ",((nftOwnerships[_nftHash].droptime-block.timestamp)/86400)," Days, ",((nftOwnerships[_nftHash].droptime-block.timestamp)/3600)," Hours,", ((nftOwnerships[_nftHash].droptime-block.timestamp)/60)," Minutes, and ", ((nftOwnerships[_nftHash].droptime-block.timestamp))," Seconds left.")));
+    function drop(uint256 _dropHash) public {
+        require(
+            nftOwnerships[_dropHash].dropTime <= block.timestamp,
```
Comment (JaPilgrim) [inline blockchain/contracts/HelloWorld.sol:73]: here dropHash is used as index to retriev Drop time. Results in Requirement not working, unless nftHash is used as Parameter. (Rest of function works as supposed, as anyways all available NFTs are to be dropped)

File: blockchain/contracts/HelloWorld.sol [hunk 3]
```diff
@@ -1,52 +1,124 @@
...
+        );
+        require(
+            nftOwnerships[_nftHash].reservedFor == msg.sender,
+            "This NFT wasn't assigned to you"
+        );
+        // string(abi.encodePacked("Drop has not started yet! ",((nftOwnerships[_nftHash].droptime-block.timestamp)/86400)," Days, ",((nftOwnerships[_nftHash].droptime-block.timestamp)/3600)," Hours,", ((nftOwnerships[_nftHash].droptime-block.timestamp)/60)," Minutes, and ", ((nftOwnerships[_nftHash].droptime-block.timestamp))," Seconds left.")));
         _price = msg.value;
         nftOwner.transfer(_price);
-        nftOwnerships[_nftHash].owner = buyer;        
+        nftOwnerships[_nftHash].owner = buyer;
```
Comment (JaPilgrim) [inline blockchain/contracts/HelloWorld.sol:101]: Error occured while Testing: 
If I try to buy the NFT for which I was "reservedFor", transaction works, but the new "owner" is set to 0x0...0.

**Review by JaPilgrim — CHANGES_REQUESTED** [review]
Otherwise works perfectly (Y)

**Comment by JaPilgrim** [comment]
And sorry for the late review while still saying I would be available for a review this day

File: blockchain/contracts/HelloWorld.sol (hunk 3 above)
Comment (BerinikeTech) [reply blockchain/contracts/HelloWorld.sol:101]: This is a remix specific issue, can't do anything about it :/

File: blockchain/contracts/HelloWorld.sol (hunk 1 above)
Comment (BerinikeTech) [reply blockchain/contracts/HelloWorld.sol:55]: What's the issue with that?

File: blockchain/contracts/HelloWorld.sol (hunk 2 above)
Comment (BerinikeTech) [reply blockchain/contracts/HelloWorld.sol:73]: nftHash is supposed to be used as a Parameter here to execute the correct drop

Comment (BerinikeTech) [reply blockchain/contracts/HelloWorld.sol:73]: What do you mean by "Results in Requirement not working, unless nftHash is used as parameter"?

File: blockchain/contracts/HelloWorld.sol (hunk 3 above)
Comment (BerinikeTech) [reply blockchain/contracts/HelloWorld.sol:101]: image
found a solution :)

File: blockchain/contracts/HelloWorld.sol (hunk 2 above)
Comment (BerinikeTech) [reply blockchain/contracts/HelloWorld.sol]: resolved with the newest commit :) I got rid of the dropHash altogether since we don't need it atm

**Comment by JaPilgrim** [comment]
Its all working now!

**Review by JaPilgrim — APPROVED** [review]

**Comment by JaPilgrim** [comment]
All working

**Pull request merged**